src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java

```diff
@@ -1222,242 +1227,322 @@ public ContentBuilder add(CharSequence text) {
-                                text = redirectRelativeLinks(element, tt);
-                            }
-                            content.add(textCleanup(text, isLastNode));
-                        } else if (dt instanceof DocRootTree drt) {
-                            // defer until we see what, if anything, follows this node
-                            pendingDocRoot = drt;
-                        } else {
-                            dt.accept(this, content);
-                        }
-                        first = false;
-                    }
-                    if (pendingDocRoot != null) {
-                        pendingDocRoot.accept(this, content);
-                    }
+        if (useMarkdown) {
+            markdownHandler.addContent(result);
+        }
 
-                    content.add(quote);
-                    return false;
-                }
+        return result;
+    }
 
-                @Override
-                public Boolean visitComment(CommentTree node, Content content) {
-                    content.add(RawHtml.comment(node.getBody()));
-                    return false;
+    private class MarkdownHandler {
+        private static final char FFFC = '\uFFFC'; // Unicode Object Replacement Character
+        StringBuilder markdownInput = new StringBuilder() ;
+        ArrayList<Content> fffcObjects = new ArrayList<>();
+
+        boolean handle(DocTree tree, InlineVisitor visitor) {
+            boolean allDone;
+            if (tree instanceof MarkdownTree t) {
+                String code = t.getCode();
+                // handle the (unlikely) case of FFFC characters existing in the code
+                int start = 0;
+                int pos;
+                while ((pos = code.indexOf(FFFC, start)) != -1) {
+                    markdownInput.append(code.substring(start, pos));
+                    markdownInput.append(FFFC);
+                    fffcObjects.add(Text.of(String.valueOf(FFFC)));
+                    start = pos + 1;
                 }
+                markdownInput.append(code.substring(start));
+                allDone = false;
+            } else {
+                Content embeddedContent = new ContentBuilder();
+                allDone = visitor.visit(tree, embeddedContent);
+                fffcObjects.add(embeddedContent);
+                markdownInput.append(FFFC);
+            }
+            return allDone;
+        }
 
-                @Override
-                public Boolean visitDocRoot(DocRootTree node, Content content) {
-                    content.add(getInlineTagOutput(element, node, context));
-                    return false;
-                }
+        void addContent(Content result) {
+            Parser parser = Parser.builder().build();
+            Node document = parser.parse(markdownInput.toString());
+            HtmlRenderer renderer = HtmlRenderer.builder().build();
+            String markdownOutput = renderer.render(document);
```

Comment:
Seems like a good idea if all works as might be expected.

Reply:
There was no perception that the Markdown machinery was particularly onerous, even in play scenarios in which all comments were treated as Markdown. Nevertheless, it seems reasonable to cache them within `MarkdownHandler`; less so in the enclosing `HtmlDocletWriter`.  It may be the case down the road that we want to configure the builders with settings that may vary between comments, so `MarkdownHandler` seems a good choice,